### PR TITLE
ThreadSafeWeakPtr should be able to be safely created during object destructor

### PR DESCRIPTION
**Pull request by achristensen07**
#### 522fb3fb9f3b48d6d23181470190a899dd2cf1d0
<pre>
ThreadSafeWeakPtr should be able to be safely created during object destructor
<a href="https://bugs.webkit.org/show_bug.cgi?id=258324">https://bugs.webkit.org/show_bug.cgi?id=258324</a>

Reviewed by Chris Dumez.

This requires a RefPtr to use traits to make it null if weakRef is called after the object&apos;s destructor
has been scheduled.  Otherwise, we get a non-null RefPtr to the control block.

* Source/WTF/wtf/CompactRefPtrTuple.h:
* Source/WTF/wtf/HashFunctions.h:
* Source/WTF/wtf/HashTraits.h:
(WTF::HashTraits&lt;RefPtr&lt;P&gt;&gt;::emptyValue): Deleted.
(WTF::HashTraits&lt;RefPtr&lt;P&gt;&gt;::peek): Deleted.
(WTF::HashTraits&lt;RefPtr&lt;P&gt;&gt;::customDeleteBucket): Deleted.
* Source/WTF/wtf/RefPtr.h:
(WTF::DefaultRefDerefTraits::refIfNotNull):
(WTF::DefaultRefDerefTraits::derefIfNotNull):
(WTF::RefPtr::RefPtr):
* Source/WTF/wtf/ThreadSafeWeakHashSet.h:
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeWeakPtrControlBlock::weakRef):
(WTF::ThreadSafeWeakPtrControlBlock::weakDeref):
(WTF::ThreadSafeWeakPtrControlBlockRefDerefTraits::refIfNotNull):
(WTF::ThreadSafeWeakPtrControlBlockRefDerefTraits::derefIfNotNull):
(WTF::ThreadSafeWeakPtrControlBlock::ref const): Deleted.
(WTF::ThreadSafeWeakPtrControlBlock::deref const): Deleted.
* Source/WebCore/Modules/webaudio/AudioNode.h:
(WebCore::AudioNodeConnectionRefDerefTraits::refIfNotNull):
(WebCore::AudioNodeConnectionRefDerefTraits::derefIfNotNull):
* Source/WebCore/PAL/pal/graphics/WebGPU/Impl/WebGPUPtr.h:
(PAL::WebGPU::release):
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/265344@main">https://commits.webkit.org/265344@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/011a3d751ca125b42cacdef60ad91bc6c032133f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/10674 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/10898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11187 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/12323 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10236 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/10689 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13269 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/10868 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13153 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/10834 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/11752 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/8972 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/12726 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9046 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/9631 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/16892 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/9032 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10116 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/9782 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13035 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/10127 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10248 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8342 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/10791 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/9408 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/2912 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/13680 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/11092 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/1193 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10110 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/2717 "Passed tests") |

### CODE REVIEW

**Review by cdumez — APPROVED** [review]
LGTM as long as the bots are happy.

**Review by fujii — COMMENTED** [review]
I don't understand what problem is fixed by this patch. It should not be a problem that "we get a non-null RefPtr to the control block". I confirmed `WTF_ThreadSafeWeakPtr.WeakRefInDestructor` passes without your change.

**Comment by fujii** [comment]
> > I don't understand what problem is fixed by this patch. It should not be a problem that "we get a non-null RefPtr to the control block". I confirmed `WTF_ThreadSafeWeakPtr.WeakRefInDestructor` passes without your change.
> 
> Did you try an ASAN build?

Ah, I get the idea. The control block is destroyed after destroying the object without re-checking weak ref-counter.
https://github.com/WebKit/WebKit/blob/7c8d3655b4cb510d9875cd000b0b791a1bfa67ed/Source/WTF/wtf/ThreadSafeWeakPtr.h#L85-L87
How about an idea of checking the weak ref-counter after destroying the object.

**Comment by achristensen07** [comment]
Fujii, I didn't see your comment before merging.  I think checking the weak reference count before destroying the control block would've been another valid way of fixing this bug.

**Comment by achristensen07** [comment]
Actually, I think that wouldn't work because you'd have to get the lock to inspect the weak reference count, and if it were zero you'd delete the control block but there may be another thread waiting on the lock which would then UAF the lock.

**Pull request merged**
